src/manager/node.c

```diff
@@ -206,6 +206,14 @@ void node_unref(Node *node) {
         free(node);
 }
 
+void node_shutdown(Node *node) {
```

Comment:
```suggestion
void node_cancel_outstanding_requests(Node *node) {
```
`shutdown` implies that it is doing a lot more, but its only canceling outstanding requests. Not sure if the proposed name is better, though.

Reply:
This was the name suggested by @alexlarsson 
Also, cancelling these requests really serve the purpose of graceful shutdown, and that is what we really are doing here

Reply:
I think shutdown is a better word. Its about what the meaning of the operation is, i.e. "prepare for a shutdown", not about what it is currently doing, which is more of an internal thing.

Tomorrow we may notice that shutdown needs to do more things, then we just add it to shutdown().

Reply:
Sometimes this operation is also called dispose, but I don't think the exact name is all that important. Its just a two phase object destruction.

Reply:
Ah ok, I didn't know about the actual purpose of this function and assumed it based on the current implementation. That makes sense. 

When do we decide to move something to `node_shutdown`? For example, in `node_unref` we are also removing all proxy monitors. So couldn't we just define a weak dependency in the `AgentRequest` on the node (similar to the proxy) and take care of it in the unref?

Reply:
Two way shutdown is a standard way to handle object finalization in C when there are risks for reference cycles.

Basically, "unref" of an object means "i don't personally need this anymore". But sometimes you want something more final than that, more like "this object should never be used anymore". In the later case you can't just jank it away from other owners of refs, as those would then reference freed memory, so instead you introduce a new state where something is "shut down, but memory is still valid". Such objects should not allow new operations on the, and all internal outstanding references should be broken to avoid any problems with circular dependencies.

In this particular case we need to shutdown the node, because we're exiting the app, and there may be references to the request objects that are external to the app (i.e. outstanding dbus calls) that make the agent requests not be cleanly freed.

But yes, you're right, one approach for ownership is to use weak references, which in this case would mean make the AgentRequests not reference the Node and have the Node finalizer just unref outstanding requests. In practice, this will work in current bluechi, because we will only do the final unref of the node when exiting the program. But, as a general principle, if someone calls an async operation like node_method_list_units() that expresses an interest in the result, and you want a callback eventually, even if you don't necessarily need to keep the node object reference around. This won't happen if the request doesn't ref the node.